lib/puppet/pops/evaluator/runtime3_converter.rb

```diff
@@ -51,98 +52,75 @@ def map_args(args, scope, undef_value)
   # @param o [Object]The value to convert
   # @param scope [Puppet::Parser::Scope] The scope to use when converting
   # @param undef_value [Object] The value that nil is converted to
+  # @param allow_extended [Boolean] Allow extended values
   # @return [Object] The converted value
   #
   def convert(o, scope, undef_value)
     @convert_visitor.visit_this_2(self, o, scope, undef_value)
   end
 
   def convert_NilClass(o, scope, undef_value)
-    undef_value
-  end
-
-  def convert2_NilClass(o, scope, undef_value)
-    :undef
+    @inner ? :undef : undef_value
   end
 
   def convert_String(o, scope, undef_value)
     # although wasteful, needed because user code may mutate these strings in Resources
     o.frozen? ? o.dup : o
   end
-  alias convert2_String :convert_String
 
   def convert_Object(o, scope, undef_value)
     o
   end
-  alias :convert2_Object :convert_Object
 
   def convert_Array(o, scope, undef_value)
-    o.map {|x| convert2(x, scope, undef_value) }
+    ic = @inner_converter
+    o.map {|x| ic.convert(x, scope, undef_value) }
   end
-  alias :convert2_Array :convert_Array
 
   def convert_Hash(o, scope, undef_value)
     result = {}
-    o.each {|k,v| result[convert2(k, scope, undef_value)] = convert2(v, scope, undef_value) }
+    ic = @inner_converter
+    o.each {|k,v| result[ic.convert(k, scope, undef_value)] = ic.convert(v, scope, undef_value) }
     result
   end
-  alias :convert2_Hash :convert_Hash
 
   def convert_Iterator(o, scope, undef_value)
     raise Puppet::Error, 'Use of an Iterator is not supported here'
   end
-  alias :convert2_Iterator :convert_Iterator
 
   def convert_Regexp(o, scope, undef_value)
     # Puppet 3x cannot handle parameter values that are reqular expressions. Turn into regexp string in
-    # source form
+    # source form unless allow_extended
```

Comment:
Part of another solution attempt. Should have been removed (and is removed now).